### PR TITLE
conversion API w/ some initial converters

### PR DESCRIPTION
**Pull request by jdef**
* generic conversion API
* network conversion
* envvar conversion
* internal Network object
* better network API validation
* pod conversion
* container conversion

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala [hunk 1]
```diff
@@ -33,10 +33,23 @@ trait PodsValidation {
...
+  } and isTrue[Seq[Network]]("Duplicate networks are not allowed") { nets =>
+    // unnamed CT nets pick up the default virtual net name
+    val unnamedAtMostOnce = nets.count { n => n.name.isEmpty && n.mode == NetworkMode.Container } < 2
     val realNamesAtMostOnce: Boolean = !nets.flatMap(_.name).groupBy(name => name).exists(_._2.size > 1)
     unnamedAtMostOnce && realNamesAtMostOnce
+  } and isTrue[Seq[Network]]("May specify a single host network, or 1-to-n container networks, but not both") { nets =>
+    val countsByMode = nets.groupBy { net => net.mode }.mapValues(_.size)
+    val hostNetworks = countsByMode.get(NetworkMode.Host).getOrElse(0)
+    val containerNetworks = countsByMode.get(NetworkMode.Container).getOrElse(0)
+    (hostNetworks == 1 && containerNetworks == 0) || (hostNetworks == 0 && containerNetworks > 0)
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]: if you specify no host network and no container network, this will fail

File: src/main/scala/mesosphere/marathon/api/v2/PodsResource.scala [hunk 2]
```diff
@@ -40,22 +40,7 @@ class PodsResource @Inject() (
...
-        } else {
-          network
-        }
-      }
-      pod.copy(networks = networks)
-    } else {
-      pod
-    }
-  }
+  private def normalize(pod: PodDefinition): PodDefinition = identity(pod)
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/api/v2/PodsResource.scala]: why do we need normalize, if it calls identity?

File: src/main/scala/mesosphere/marathon/core/pod/Network.scala [hunk 3]
```diff
@@ -0,0 +1,37 @@
+package mesosphere.marathon.core.pod
+
+import mesosphere.marathon.raml.{ KVLabels, NetworkMode, Network => RAMLNetwork }
+
+/**
+  * Network declared by a [[PodDefinition]].
+  */
+sealed trait Network extends Product with Serializable {
+  def toAPIObject: RAMLNetwork
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/pod/Network.scala]: We start adding api level code into the core.
Wdyt putting this logic into a conversion class/package under api?

**Comment by aquamatthias** [comment]
@jdef Please rebase

**Comment by jdef** [comment]
discussed with MV, going to create a conversions package and migrate the logic there. flipping this back into WIP

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala (hunk 1 above)
Comment (jdef) [inline src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]: that's right. I guess we should support "no networking"?

File: src/main/scala/mesosphere/marathon/api/v2/PodsResource.scala (hunk 2 above)
Comment (jdef) [inline src/main/scala/mesosphere/marathon/api/v2/PodsResource.scala]: not sure, @jasongilanfarr added these

File: src/main/scala/mesosphere/marathon/core/pod/Network.scala (hunk 3 above)
Comment (jdef) [inline src/main/scala/mesosphere/marathon/core/pod/Network.scala]: WIP

File: src/main/scala/mesosphere/marathon/api/v2/PodsResource.scala (hunk 2 above)
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/api/v2/PodsResource.scala]: It'll make the conversion to akka http easier. Trying to get us thinking in a pattern ;)

File: src/main/scala/mesosphere/marathon/api/v2/conversion/Converter.scala [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package mesosphere.marathon.api.v2.conversion
+
+trait Converter[A,B] {
```
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/api/v2/conversion/Converter.scala]:
```
trait Reads[A] {
   def read(a: RamlType): A
}

trait Writes[A] {
   def write(a: A): RamlType
}
```

etc.

File: src/main/scala/mesosphere/marathon/api/v2/conversion/EnvVarConversion.scala [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package mesosphere.marathon.api.v2.conversion
+
+import mesosphere.marathon.raml.EnvVars
+import mesosphere.marathon.{ raml, state }
+
+import scala.collection.immutable.Map
+
+trait EnvVarConversion {
+
+  implicit val fromAPIObjectToEnvironment = Converter { envVars: EnvVars =>
```
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/api/v2/conversion/EnvVarConversion.scala]: naming: ramlToEnvVar. add type annotation.

File: src/main/scala/mesosphere/marathon/api/v2/conversion/EnvVarConversion.scala [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    def toEnvValue(envVarValue: raml.EnvVarValueOrSecret): state.EnvVarValue = envVarValue match {
+      case raml.EnvVarValue(value) => state.EnvVarString(value)
+      case raml.EnvVarSecretRef(secret) => state.EnvVarSecretRef(secret)
+    }
+    if(envVars.values.isEmpty) None else Some(envVars.values.map{ case (k, v) => k -> toEnvValue(v) })
+  }
+
+  implicit val fromEnvironmentToAPIObject = Converter { envMap: Map[String, state.EnvVarValue] =>
```
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/api/v2/conversion/EnvVarConversion.scala]: envVarToRaml, add type annotation

File: src/main/scala/mesosphere/marathon/api/v2/conversion/HealthCheckConversion.scala [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package mesosphere.marathon.api.v2.conversion
+
+import mesosphere.marathon.core.health.HealthCheck
+import mesosphere.marathon.raml.{ HealthCheck => RAMLHealthCheck }
+
+trait HealthCheckConversion {
```
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/api/v2/conversion/HealthCheckConversion.scala]: see above...

File: src/main/scala/mesosphere/marathon/api/v2/conversion/Converter.scala [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+  def convert(a: A): Option[B]
```
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/api/v2/conversion/Converter.scala]: Should just be "B"

File: src/main/scala/mesosphere/marathon/api/v2/conversion/Converter.scala [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+}
+
+object Converter {
+  def apply[A,B](a: A)(implicit c: Converter[A,B]): Option[B] = c.convert(a)
```
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/api/v2/conversion/Converter.scala]: No option.

File: src/main/scala/mesosphere/marathon/api/v2/conversion/Converter.scala [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+
+  def apply[A,B](f: (A) => Option[B]): Converter[A,B] = new Converter[A,B] {
+    override def convert(a: A): Option[B] = f(a)
```
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/api/v2/conversion/Converter.scala]: No option

File: src/main/scala/mesosphere/marathon/api/v2/conversion/Converter.scala (hunk 4 above)
Comment (jdef) [inline src/main/scala/mesosphere/marathon/api/v2/conversion/Converter.scala]: if only we had a RamlType :)

File: src/main/scala/mesosphere/marathon/api/v2/conversion/Converter.scala (hunk 8 above)
Comment (jdef) [inline src/main/scala/mesosphere/marathon/api/v2/conversion/Converter.scala]: done

File: src/main/scala/mesosphere/marathon/api/v2/conversion/Converter.scala (hunk 9 above)
Comment (jdef) [inline src/main/scala/mesosphere/marathon/api/v2/conversion/Converter.scala]: done

File: src/main/scala/mesosphere/marathon/api/v2/conversion/Converter.scala (hunk 10 above)
Comment (jdef) [inline src/main/scala/mesosphere/marathon/api/v2/conversion/Converter.scala]: done

File: src/main/scala/mesosphere/marathon/api/v2/conversion/ContainerConversion.scala [hunk 11]
```diff
@@ -0,0 +1,42 @@
+package mesosphere.marathon.api.v2.conversion
```
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/api/v2/conversion/ContainerConversion.scala]: I'm kind of thinking this should be in mesosphere/marathon/raml (or mesosphere/marathon/core/raml)

**Comment by jdef** [comment]
of the two options mentioned, core/raml preferred. the package that
contains generated code (mesosphere/marathon/raml) shouldn't have much else
inside of it.

**Comment by jasongilanfarr** [comment]
I'm ok with either and actually having the conversion in the same package seems correct to me.

File: src/main/scala/mesosphere/marathon/core/pod/PodDefinition.scala [hunk 12]
```diff
@@ -88,142 +88,27 @@ case class PodDefinition(
...
-      constraints = constraints,
-      version = podDef.version.fold(Timestamp.now())(Timestamp(_)),
-      podVolumes = podDef.volumes,
-      networks = networks,
-      backoffStrategy = backoffStrategy,
-      upgradeStrategy = upgradeStrategy
-    )
-  }
-  //scalastyle:on
+  def apply(podDef: Pod): PodDefinition = Converter[Pod,PodDefinition](podDef)
```
Comment (jdef) [inline src/main/scala/mesosphere/marathon/core/pod/PodDefinition.scala]: shouldn't need explicit `Converter[...]` typing here

File: src/main/scala/mesosphere/marathon/raml/NetworkConversion.scala [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+
+import mesosphere.marathon.core.pod
+
+trait NetworkConversion {
+  implicit val networkRamlReader: Reads[Network, pod.Network] =
+    Reads { raml =>
+      raml.mode match {
+        case NetworkMode.Host => pod.HostNetwork
+        case NetworkMode.Container => pod.ContainerNetwork(
+          // TODO(PODS): shouldn't this be caught by validation?
```
Comment (jdef) [inline src/main/scala/mesosphere/marathon/raml/NetworkConversion.scala]: yes, but how else would you handle this (unexpected) case?

Comment (jdef) [inline src/main/scala/mesosphere/marathon/raml/NetworkConversion.scala]: actually, handled by normalization

File: scalastyle-config.xml [hunk 14]
```diff
@@ -1,127 +0,0 @@
-<scalastyle commentFilter="enabled">
```
Comment (jdef) [inline scalastyle-config.xml]: why is this file removed as part of this PR?

File: src/main/scala/mesosphere/marathon/raml/NetworkConversion.scala (hunk 13 above)
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/raml/NetworkConversion.scala]: True on both counts

File: scalastyle-config.xml (hunk 14 above)
Comment (jasongilanfarr) [inline scalastyle-config.xml]: I was annoyed with the formatted violating the long lines by 1 character and it doesn't exist on the branch.

**Comment by jdef** [comment]
since the raml types are all specific to v2 a better home for the conversion stuff is probably under api/v2 instead of core

**Comment by jasongilanfarr** [comment]
Don't agree at all. We currently only have v2 defined in raml. Please leave
it where it is.

**Comment by jdef** [comment]
I left the location alone. Discussion tabled.

**Comment by jasongilanfarr** [comment]
Long term, I think the raml converter itself should stay in the package and
the generator should support package annotations (and imports) so we can
organize accordingly.

**Comment by jdef** [comment]
That's fine. There's probably a limit to what we'll actually want to
automated in terms of converter generation: rewards we reap vs. time
invested (and RAML readability). I suspect we'll land on a solution that
generates 90 - 95% of the converter code, keeps the RAML readable, and
still requires some manual bits to fill in the gaps.
